test/parallel/test-webstreams-compose.js

```diff
@@ -0,0 +1,364 @@
+      })
+    }),
+    async function*(source) {
+      for await (const chunk of source) {
+        yield chunk;
+      }
+    },
+    new Writable({
+      write: common.mustCall((chunk, encoding, callback) => {
+        res += chunk;
+        callback(null);
+      })
+    })
+  )
+    .end('asd')
+    .on('finish', common.mustCall(() => {
+      assert.strictEqual(res, 'ASD');
+    }));
+}
+
+{
+  let res = '';
+  compose(
+    new Transform({
+      transform: common.mustCall((chunk, encoding, callback) => {
+        callback(null, chunk.toString().toUpperCase());
+      })
+    }),
+    async function*(source) {
+      for await (const chunk of source) {
+        yield chunk;
+      }
+    },
+    new WritableStream({
+      write: common.mustCall((chunk) => {
+        res += chunk;
+      })
+    })
+  )
+  .end('asd')
+  .on('finish', common.mustCall(() => {
+    assert.strictEqual(res, 'ASD');
+  }));
+}
+
+{
+  let res = '';
+  compose(
+    new TransformStream({
+      transform: common.mustCall((chunk, controller) => {
+        controller.enqueue(chunk.toString().toUpperCase());
+      })
+    }),
+    async function*(source) {
+      for await (const chunk of source) {
+        yield chunk;
+      }
+    },
+    new WritableStream({
+      write: common.mustCall((chunk) => {
+        res += chunk;
+      })
+    })
+  )
+  .end('asd')
+  .on('finish', common.mustCall(() => {
+    assert.strictEqual(res, 'ASD');
+  }));
+}
+
+{
+  let res = '';
+  compose(
+    new TransformStream({
+      transform: common.mustCall((chunk, controller) => {
+        controller.enqueue(chunk.toString().toUpperCase());
+      })
+    }),
+    async function*(source) {
+      for await (const chunk of source) {
+        yield chunk;
+      }
+    },
+    async function(source) {
+      for await (const chunk of source) {
+        res += chunk;
+      }
+    }
+  )
+  .end('asd')
+  .on('finish', common.mustCall(() => {
+    assert.strictEqual(res, 'ASD');
+  }));
+}
+
+{
+
+  compose(
+    new TransformStream({
+      transform: common.mustCall((chunk, controller) => {
+        controller.error(new Error('asd'));
+      })
+    }),
+    new TransformStream({
+      transform: common.mustNotCall()
+    })
+  )
+  // .on('data', common.mustNotCall())
+  .on('end', common.mustNotCall())
+  .on('error', (err) => {
+    assert.strictEqual(err?.message, 'asd');
+  })
+  .end('xyz');
+}
+
+{
+  const _err = new Error('asd');
+
+  compose(
+    new TransformStream({
+      transform: common.mustCall((chunk, controller) => {
+        controller.enqueue(chunk);
+      })
+    }),
+    async function*(source) { // eslint-disable-line require-yield
+      let tmp = '';
+      for await (const chunk of source) {
+        tmp += chunk;
+        throw _err;
+      }
+      return tmp;
+    },
+    new TransformStream({
+      transform: common.mustNotCall()
+    })
+  )
+  // .on('data', common.mustNotCall())
```

Comment:
One small issue I am facing putting a on 'data' listener breaks the test with the following error message

```
node:internal/process/promises:289
            triggerUncaughtException(err, true /* fromPromise */);
            ^

Error: asd
    at Object.<anonymous> (/Users/debadreechatterjee/Documents/Personal/node/test/parallel/test-webstreams-compose.js:322:26)
    at Object.transform (/Users/debadreechatterjee/Documents/Personal/node/test/common/index.js:448:15)
    at ensureIsPromise (node:internal/webstreams/util:182:19)
    at transformStreamDefaultControllerPerformTransform (node:internal/webstreams/transformstream:509:18)
    at transformStreamDefaultSinkWriteAlgorithm (node:internal/webstreams/transformstream:559:10)
    at Object.write (node:internal/webstreams/transformstream:364:14)
    at ensureIsPromise (node:internal/webstreams/util:182:19)
    at writableStreamDefaultControllerProcessWrite (node:internal/webstreams/writablestream:1112:5)
    at writableStreamDefaultControllerAdvanceQueueIfNeeded (node:internal/webstreams/writablestream:1227:5)
    at writableStreamDefaultControllerWrite (node:internal/webstreams/writablestream:1101:3)
```

but the data listener isn't actually called so could this be an issue with `common.mustNotCall` ?